### ISSUE TITLE
Multiclass classification

### ISSUE THREAD

**Issue by hasan-sayeed** [anthony-wang/CrabNet#10]
Is there an easy way to modify crabnet for a multi-class classification problem? I can see there is a built-in function for binary classification. I mean, is there any way other than introducing my own custom function (let's say which uses CrossEntropyLoss) for the task?

**Comment by anthony-wang** [comment]
Hey, thanks for the question. Yeah I believe the simplest way is to introduce your own changes to the CrabNet architecture to satisfy multi-class classification--it should be relatively straightforward to do so!

You can also check out this paper, where they habe modified CrabNet, Roost and ElemNet to address multi-class prediction problems: https://arxiv.org/abs/2106.02225.

**Comment by sgbaird** [comment]
Hmm.. in the repo linked on arxiv, "CrabNet" returned 0 search results.

**Comment by sgbaird** [comment]
@hasan-sayeed, looks like MODNet implemented multi-class classification. Sounds like they just replaced a softmax activation function with a sigmoid. See https://github.com/ppdebreuck/modnet/pull/70/files. In particular, see the changes in `modnet/models/vanilla.py`. They use `tf` and `keras` instead of `pytorch`, and pass in something to `tf.keras.layers.Dense` as shown [here](https://github.com/ppdebreuck/modnet/blob/cf35c995596c31104ceecb0d826bd10824455803/modnet/models/vanilla.py#L194-L198). I'd have to dig a bit deeper (in reality, shallower in terms of MODNet's stack trace) to see how they use the sigmoid, but I imagine that for 20 possible labels, they just divide the sigmoid into 20 equal segments and find which bin it fits into. Let me know if you need help with this. If you open up an issue with MODNet asking about it, they would probably help out, too.

**Comment by sgbaird** [comment]
Also, checking back on the linked paper - while they do say they modified `CrabNet` to account for multi-target predictions (e.g. both bulk modulus and shear modulus for example) as well as a transfer learning functionality within `CrabNet`, I couldn't find anywhere in the paper that talked about them implementing multi-class classification. Incidentally, they do mention that "the attention mechanism is the crux behind many state-of-the-art sequence-to-sequence models used in machine translation and language processing[39] and it has recently shown good results on multi-label classification[40]".

Since these seem like useful additions, I figured it would be worth asking for their implementations. https://github.com/gomes-lab/H-CLMP/issues/1

**Comment by sgbaird** [comment]
Hmm... is it as simple as changing `out_dims` to be `(3, `n_classes`)` ([inspiration](https://towardsdatascience.com/pytorch-tabular-multiclass-classification-9f8211a123ab)):
https://github.com/anthony-wang/CrabNet/blob/a5be06f75b3e524a4485fe00139c5c06aea9cfe5/crabnet/kingcrab.py#L191-L197
and then just fiddle with:
https://github.com/anthony-wang/CrabNet/blob/a5be06f75b3e524a4485fe00139c5c06aea9cfe5/crabnet/model.py#L100-L104 and anywhere else along the way that assumed a scalar dimension? Seems like the multi-target aspect would follow essentially the same process. I've also been seeing a lot that says to use `nn.CrossEntropyLoss` instead of `nn.BCEWithLogitsLoss` since the former is better for strictly multi-class problems but the latter is better for multi-label problems (multiple labels can be assigned to a single datapoint).

Might still be a bit tricky to implement, but I think the approach is solid.
